src/adapter_gen.rs

```diff
@@ -50,19 +50,19 @@ pub trait WS28xxAdapter {
 /// hardware interface to your WS28xx LEDs. *Handle this as something like an abstract class
 /// for concrete implementations!* This works in `#[no-std]`-environments.
 pub struct WS28xxGenAdapter {
-    hw: Box<dyn HardwareDev>,
+    hw: Box<dyn HardwareDev<Error=String>>,
```

Comment:
It's a little bizarre that the original library used dynamic dispatch polymorphism (the `Box<dyn ...>` stuff) for implementing it's hardware devices - you should be able to replace these with `impl HardwareDev<Error=...>` and lose the additional dependency on `alloc` which is caused by `Box`, plus glean back some runtime performance. Further note that some platforms (notably arduino last I checked) do not properly support dynamic dispatch with rust at the moment.